File: concrete/src/main/java/pcb/uwu/amount/finance/EUR.java
package pcb.uwu.amount.finance;

import org.jetbrains.annotations.NotNull;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.unit.finance.CurrencyUnit;
import pcb.uwu.utils.UnitAmountUtils;

import java.math.BigDecimal;

import static pcb.uwu.utils.UnitAmountUtils.dividedByScalar;
import static pcb.uwu.utils.UnitAmountUtils.minusAmount;
import static pcb.uwu.utils.UnitAmountUtils.plusAmount;

public class EUR extends Currency {

	public static final CurrencyUnit CURRENCY = CurrencyUnit.of("EUR");

	// region constructor

	public EUR(Number value) {
		super(value, CURRENCY);
	}

	public EUR(String value) {
		super(value, CURRENCY);
	}

	public EUR(BigDecimal value) {
		super(value, CURRENCY);
	}

	public EUR(BigDecimalAmount amount) {
		super(amount, CURRENCY);
	}

	// endregion

	// region implement UnitAmount

	public EUR plus(EUR other) {
		return new EUR(getAmount().plus(other.getAmount().getValue()));
	}

	public EUR minus(EUR other) {
		return new EUR(getAmount().minus(other.getAmount().getValue()));
	}

	@NotNull
	@Override
	public EUR plus(@NotNull UnitAmount<CurrencyUnit> currency) {
		return new EUR(plusAmount(this, currency));
	}

	@NotNull
	@Override
	public Currency minus(@NotNull UnitAmount<CurrencyUnit> currency) {
		return new AUD(minusAmount(this, currency));
	}

	@NotNull
	@Override
	public EUR times(@NotNull Number number) {
		return new EUR(UnitAmountUtils.times(this, number));
	}

	@NotNull
	@Override
	public EUR div(@NotNull Number number) {
		return new EUR(dividedByScalar(this, number));
	}

	// endregion
}

File: concrete/src/main/java/pcb/uwu/amount/base/currency/EUR.kt
package pcb.uwu.amount.base.currency

import pcb.uwu.amount.base.Currency
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.CurrencyUnit
import pcb.uwu.unit.base.CurrencyUnit.Companion.EUR

class EUR : Currency
{
    @JvmOverloads
    constructor(amount: Number, magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, EUR)

    @JvmOverloads
    constructor(value: String, magnitude: Magnitude = NATURAL)
            : super(value, magnitude, EUR)

    // region UnitAmount

    override fun plus(currency: UnitAmount<CurrencyUnit>) =
        EUR(this.amount + (currency to EUR).amount)

    override fun minus(currency: UnitAmount<CurrencyUnit>) =
        EUR(this.amount - (currency to EUR).amount)

    override fun times(scalar: Number) =
        EUR(this.amount * scalar)

    override fun div(scalar: Number) =
        EUR(this.amount / scalar)

    // endregion
}